### PR TITLE
Perf doc v1

### PR DESCRIPTION
**Pull request by pevma**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

- https://redmine.openinfosecfoundation.org/issues/2839

Describe changes:
- doc: Update tuning considerations doc
- doc: Update high performance config doc

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: doc/userguide/performance/high-performance-config.rst [hunk 1]
```diff
@@ -1,19 +1,341 @@
...
+ ifconfig eth1 down
+ /usr/local/sbin/ethtool -L eth1 combined 16
+ /usr/local/sbin/ethtool -K eth1 rxhash on
+ /usr/local/sbin/ethtool -K eth1 ntuple on
+ ifconfig eth1 up
+ /usr/local/sbin/ethtool -X eth1 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 16
+ /usr/local/sbin/ethtool -A eth1 rx off tx off
+ /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+ /usr/local/sbin/ethtool -G eth1 rx 1024
+
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: maybe explain the different settings a bit more or reference to the manpage of ethtool and add just a small reasoning behind the settings.

File: doc/userguide/performance/high-performance-config.rst [hunk 2]
```diff
@@ -1,19 +1,341 @@
...
+::
+
+ /usr/local/sbin/ethtool -X eth1 hfunc toeplitz
+ 
+Let the NIC balance as much as possible:
+
+::
+
+ for proto in tcp4 udp4 tcp6 udp6; do
+    /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: maybe add that for some just **sd** might work better?

File: doc/userguide/performance/high-performance-config.rst [hunk 3]
```diff
@@ -1,19 +1,341 @@
...
+    #  * cluster_flow: all packets of a given flow are send to the same socket
+    #  * cluster_cpu: all packets treated in kernel by a CPU are send to the same socket
+    #  * cluster_qm: all packets linked by network card to a RSS queue are sent to the same
+    #  socket. Requires at least Linux 3.14.
+    #  * cluster_ebpf: eBPF file load balancing. See doc/userguide/capture-hardware/ebpf-xdp.rst for
+    #  more info.
+    # Recommended modes are cluster_flow on most boxes and cluster_cpu or cluster_qm on system
+    # with capture card using RSS (require cpu affinity tuning and system irq tuning)
+    cluster-type: cluster_qm
+    ...
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: I wouldn't copy the comments and just use the 3 lines, but maybe also add some more setings:

    use-mmap: yes
    mmap-locked: yes
    tpacket-v3: yes
    rollover: no
    use-emergency-flush: yes
    ring-size: ?
    block-size: ?

File: doc/userguide/performance/high-performance-config.rst [hunk 4]
```diff
@@ -1,19 +1,341 @@
...
+    ./set_irq_affinity local eth1
+    /usr/local/sbin/ethtool -X eth1 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 36
+    /usr/local/sbin/ethtool -A eth1 rx off tx off
+    /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+    /usr/local/sbin/ethtool -G eth1 rx 1024
+    for proto in tcp4 udp4 tcp6 udp6; do
+        echo "/usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn"
+        /usr/local/sbin/ethtool -N eth1 rx-flow-hash $proto sdfn
+    done
+
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: duplicates many stuff from above, maybe just a reference to most parts?

File: doc/userguide/performance/high-performance-config.rst [hunk 5]
```diff
@@ -1,19 +1,341 @@
...
+    BogoMIPS:            4589.92
+    Virtualization:      VT-x
+    L1d cache:           32K
+    L1i cache:           32K
+    L2 cache:            256K
+    L3 cache:            46080K
+    NUMA node0 CPU(s):   0-17,36-53
+    NUMA node1 CPU(s):   18-35,54-71
+
+It is recommended that 36 worker threads are used and the NIC set up could be as follows:
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: There is some discussion to keep at least the first core of a CPU free for some "management" stuff. Especially if there is not just the NIC on that node, for example also the Disks and/or the NICs used for ssh connection.

File: doc/userguide/performance/high-performance-config.rst [hunk 6]
```diff
@@ -1,19 +1,341 @@
...
+  # worker-cpu-set is used for 'worker' threads
+  #
+  # Additionally, for autofp these apply:
+  # receive-cpu-set is used for capture threads
+  # verdict-cpu-set is used for IPS verdict threads
+  #
+  cpu-affinity:
+    - management-cpu-set:
+        cpu: [ "1-10" ]  # include only these CPUs in affinity settings
+    - receive-cpu-set:
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: this is not necessary in that case right? won't hurt either.
Also would remove the comments.

File: doc/userguide/performance/high-performance-config.rst [hunk 7]
```diff
@@ -1,19 +1,341 @@
...
+    threads: 18 
+    cluster-id: 99
+    cluster-type: cluster_qm
+    defrag: no
+    use-mmap: yes
+    mmap-locked: yes
+    tpacket-v3: yes
+    ring-size: 100000
+    block-size: 1048576
+
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: maybe explain why two sections are used and maybe a bit about why you would recommend those values

File: doc/userguide/performance/high-performance-config.rst [hunk 8]
```diff
@@ -1,19 +1,341 @@
...
+    NUMA node0 CPU(s):     0-7,64-71
+    NUMA node1 CPU(s):     8-15,72-79
+    NUMA node2 CPU(s):     16-23,80-87
+    NUMA node3 CPU(s):     24-31,88-95
+    NUMA node4 CPU(s):     32-39,96-103
+    NUMA node5 CPU(s):     40-47,104-111
+    NUMA node6 CPU(s):     48-55,112-119
+    NUMA node7 CPU(s):     56-63,120-127
+
+The `ethtool`, `show_irq_affinity.sh` and `set_irq_affinity_cpulist.sh` tools are provided from the official river sources. In the
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: **d**river

File: doc/userguide/performance/high-performance-config.rst [hunk 9]
```diff
@@ -1,19 +1,341 @@
...
+    use-mmap: yes
+    mmap-locked: yes
+    tpacket-v3: yes
+    ring-size: 100000
+    block-size: 1048576
 
-Be advised, however, that this may require lots of RAM for even modestly sized rule sets.  Also be aware that having additional CPU's available provides a greater performance boost than having more RAM available.  That is, it would be better to spend money on CPU's instead of RAM when configuring a system.
 
-It may also lead to significantly longer rule loading times.
+In the example above we have 15 RSS queues pinned to cores 1-7,64-71 on NUMA node 0 and 40 worker threads using other CPUs on different NUMA nodes.
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: Can you explain this a bit more why you pinned only 15?

File: doc/userguide/performance/tuning-considerations.rst [hunk 10]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+    ....
+    # To use the ring feature of AF_PACKET, set 'use-mmap' to yes
+    use-mmap: yes
+    # Lock memory map to avoid it goes to swap. Be careful that over subscribing could lock
+    # your system
+    #mmap-locked: yes
+    # Use tpacket_v3 capture mode, only active if use-mmap is true
+    # Don't use it in IPS or TAP mode as it causes severe latency
+    tpacket-v3: yes
+
```
Comment (norg) [inline doc/userguide/performance/tuning-considerations.rst]: maybe reference this in the other part above as I was wondering why they were not included :)

File: doc/userguide/performance/tuning-considerations.rst [hunk 11]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+    ring-size: 100000
+
+it means there will be 100 000 packets allowed in each buffer of the 5 threads. If any of the buffers gets filled (for example packet processing can not keep up) that will result in packet `drop` counters increasing in the stats logs.   
+
+The memory used for those is set up and dedicated at start and is calculated roughly like so: 
+
+::
+ af-packet.threads X af-packet.ring-size X (default-packet-size + ~750 bytes)
+
+where `af-packet.threads`, `af-packet.ring-size`, `default-packet-size` are the values set in suricata.yaml. Config values for example for af-packet could be quickly displayed with on the command line as well with `suricata --dump-config |grep af-packet`.
```
Comment (norg) [inline doc/userguide/performance/tuning-considerations.rst]: the **with** after displayed should be removed

File: doc/userguide/performance/high-performance-config.rst [hunk 12]
```diff
@@ -1,19 +1,341 @@
...
+ /usr/local/sbin/ethtool -X eth1 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 16
+ /usr/local/sbin/ethtool -A eth1 rx off tx off
+ /usr/local/sbin/ethtool -C eth1 adaptive-rx off adaptive-tx off rx-usecs 125
+ /usr/local/sbin/ethtool -G eth1 rx 1024
+
+Make sure the RSS hash function is Toeplitz:
+
+::
+
+ /usr/local/sbin/ethtool -X eth1 hfunc toeplitz
```
Comment (norg) [inline doc/userguide/performance/high-performance-config.rst]: maybe add the ethtool version needed to have this feature fully supported.
Maybe also add how to check all those settings are set to make sure every command worked as expected.

File: doc/userguide/performance/tuning-considerations.rst [hunk 13]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+::
+
+ af-packet:
+  - interface: eth0
+    # Number of receive threads. "auto" uses the number of cores
+    threads: 5
+    ....
+    ....
+    # Ring size will be computed with respect to max_pending_packets and number
+    # of threads. You can set manually the ring size in number of packets by setting
```
Comment (jlucovsky) [inline doc/userguide/performance/tuning-considerations.rst]: Suggest changing sentence starting with "You can set ..." to "The ring size can be manually set; choose a value for the number of packets for the ring."

File: doc/userguide/performance/tuning-considerations.rst [hunk 14]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+    # the following value. If you are using flow cluster-type and have really network
+    # intensive single-flow you could want to set the ring-size independently of the number
+    # of threads:
+    ring-size: 100000
+
+it means there will be 100 000 packets allowed in each buffer of the 5 threads. If any of the buffers gets filled (for example packet processing can not keep up) that will result in packet `drop` counters increasing in the stats logs.   
+
+The memory used for those is set up and dedicated at start and is calculated roughly like so:
```
Comment (jlucovsky) [inline doc/userguide/performance/tuning-considerations.rst]: s/is calculated roughly like so/is calculated as follows/

File: doc/userguide/performance/tuning-considerations.rst [hunk 15]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+  inline: auto                  # auto will use inline mode in IPS mode, yes or no set it statically
+  bypass: yes
+  reassembly:
+    memcap: 256mb
+    depth: 1mb                  # reassemble 1mb into a stream
+    toserver-chunk-size: 2560
+    toclient-chunk-size: 2560
+    randomize-chunk-size: yes
+  
+inspection will be skipped when `stream.depth` of 1mb is reached for a particular flow.
```
Comment (jlucovsky) [inline doc/userguide/performance/tuning-considerations.rst]: nit: s/inspection/Inspection/

File: doc/userguide/performance/tuning-considerations.rst [hunk 16]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
+::
+
+ af-packet:
+  - interface: eth0
+    ....
+    ....
+    ....
+    # To use the ring feature of AF_PACKET, set 'use-mmap' to yes
+    use-mmap: yes
+    # Lock memory map to avoid it goes to swap. Be careful that over subscribing could lock
```
Comment (jlucovsky) [inline doc/userguide/performance/tuning-considerations.rst]: Suggest changing these sentences:
"Locak the memory msp do it's not swapped. Note that oversubscribing locked memory could adversely affect your system."

File: doc/userguide/performance/tuning-considerations.rst [hunk 17]
```diff
@@ -24,26 +28,95 @@ The detection engine tries to split out separate signatures into
...
     custom-values:
-      toclient-groups: 50
-      toserver-groups: 50
+      toclient-groups: 100
+      toserver-groups: 100
 
-In general, increasing will improve performance, but will lead to
-higher memory usage.
+In general, increasing will improve performance. It will lead to minimal increase in memory usage. 
+The default value for `toclient-groups` and `toclient-groups` with `detect.profile: high` is 75.
```
Comment (jlucovsky) [inline doc/userguide/performance/tuning-considerations.rst]: Should this be `toclient-groups` and `toserver-groups`?

File: doc/userguide/performance/high-performance-config.rst [hunk 18]
```diff
@@ -1,19 +1,341 @@
...
+          low: [ 0 ]
+          medium: [ "1" ]
+          high: [ "18-35","54-71" ]
+          default: "high"
+    #- verdict-cpu-set:
+    #    cpu: [ 0 ]
+    #    prio:
+    #      default: "high"
+
+In the af-packet section of suricata.yaml config :
```
Comment (jlucovsky) [inline doc/userguide/performance/high-performance-config.rst]: nit: space before `:`

**Comment by pevma** [comment]
Comments addressed in - https://github.com/OISF/suricata/pull/4182

**Pull request closed without merging**